libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m

```diff
@@ -416,35 +414,34 @@ - (NSString *)sfHybridViewUserAgentString
 
 - (NSURL *)frontDoorUrlWithReturnUrl:(NSString *)returnUrlString returnUrlIsEncoded:(BOOL)isEncoded createAbsUrl:(BOOL)createAbsUrl
 {
+
     // Special case: if returnUrlString itself is a frontdoor.jsp URL, parse its parameters and rebuild.
     if ([returnUrlString containsString:@"frontdoor.jsp"]) {
         return [self parseFrontDoorReturnUrlString:returnUrlString encoded:isEncoded];
     }
-    
     SFOAuthCredentials *creds = [SFUserAccountManager sharedInstance].currentUser.credentials;
     NSURL *instUrl = creds.apiUrl;
     NSString *fullReturnUrlString = returnUrlString;
-    
+
     /*
      * We need to use the absolute URL in some cases and relative URL in some
      * other cases, because of differences between instance URL and community URL.
      */
-    if (createAbsUrl && ![returnUrlString hasPrefix:@"http"]) {
+    if (createAbsUrl && ![returnUrlString containsString:@"http"]) {
```

Comment:
This was the real culprit (the difference between `hasPrefix` and `containsString`). Some URLs can be of the format `/visualforce/session?url=<blah>` and these should NOT be treated as relative URLs and prefixed by `instanceUrl`. We were doing that causing `fullReturnUrlString` a couple of lines below to become `null` and thereby `retUrl` to be `null`. However, if these URLs are passed along as-is, the page loads fine as expected post-refresh. cc: @khawkins.

Reply:
They should not be treated as relative URLs?  Why?  They are relative URLs.  Can you explain a little more?

Reply:
It should be passed along as-is for `retUrl`, but it's being prepended by `instanceUrl` incorrectly.

Reply:
But then the actual issue is that this method is being called with the wrong `createAbsUrl` value.

As mentioned earlier, this seems like a special case that's very similar to [the initial call](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/2438/files#diff-bafe72208f1ff3afa7b2024a36312efeR420) in this method.  I'd prefer to special-case it, rather than abandon the contract specified by the method.  You're going to get false negatives for things like `returnUrlString = @"/apex/httputils"` and the like.

Reply:
`parseFrontDoorReturnUrlString:encoded:` seems like exactly this case, with another URL format.  I.e. the actual return URL is embedded in the outer URL string.  The method parses that out, and runs it back through `frontDoorUrlWithReturnUrl:returnUrlIsEncoded:createAbsUrl:`.  It's not my favorite workflow (by a long stretch), but it's what makes the most sense in the current code base.

Reply:
The behavior of something like `/apex/httputils` would be the same as before, even after this fix. In fact, nothing would change except session prefix requests already embedded with a full URL. I don't view this as a special case, it has always been this way (if the value of `startURL` contains `https://`, don't touch it). The regression was changing it to "if the value of `startURL` starts with `https://`).

Reply:
I got my edge case wrong, but the problem is still there.  An absolute URL string as a return URL vs. a relative one is totally based on a token that's insufficiently determinant. For cases of createAbsUrl:

- /apex/httputils: Return URL string = /apex/httputils
- /apex/mypage: Return URL string = https://na45.salesforce.com/apex/mypage.

It's totally arbitrary, and the first case fails its mandate to create an absolute URL.

Reply:
You could update it to something like:
```obj-c
if (createAbsUrl && !([returnUrlString containsString:@"http://"] || [returnUrlString containsString:@"https://"])) {
```
Which is better, but it's still a shot in the dark about knowing what's actually going on in the data.  It may or may not make a difference, in the end, but it certainly makes it a difficult method to maintain, because it's hard to understand what's going on.